### ISSUE TITLE
Investigate feasibility of “Extract Method” IDE feature

### ISSUE THREAD

**Issue by robin-aws** [dafny-lang/dafny#1747]
This can be a cleaner, easier-to-understand solution to chop up expensive methods and improve stability. The biggest challenge is whether we can automatically determine the specification (requires, ensures, modifies, etc.) for the extracted method.

**Comment by MikaelMayer** [comment]
If you can solve this single problem, you actually solve verification stability. Here is why:: If the statement you extract is an "assert", determining which `requires` are necessary is the core of the problem.

**Comment by keyboardDrummer** [comment]
> If you can solve this single problem, you actually solve verification stability

Do you mean that you solve the problem of making a proof verify?

I think it's fine if extract method only works for proofs that are currently verifying, where it would use output from the solver to help with specifying the extracted methods clauses.

**Comment by MikaelMayer** [comment]
>Do you mean that you solve the problem of making a proof verify?

Yes, but that also applies when a proof verifies sometimes and sometimes not, due to resource count being hit.
If you know which particular assertions are used, you can provide them during refactoring and suddenly the proof would be faster.

One problem of refactoring a method is that, inside the method, everything is accessible to the verifier. If you wanted to cut a method in two, you'd need to creates an ensure and require that explicit everything that is hold to be true inside that method at the cutting point.

We can definitely help if we are only refactoring the entire branch of an if-then-else if that's the only thing of the method body, because we can add the guards as "requires" in the newly extracted methods, and keep the same "ensures".

We can also do refactoring without taking care of ensures and require clauses, or the minimal work, such as putting back the original method's ensures if the selection contains the last statement, or taking the original method's require if the selection contains the first statement, and just ask the user to actually fill in anything else.
In this sense, "refactoring" would not be "fool-proof", it's just another tool to explore with verification.
